(PDB-1313) Uniformly format json dates

rbrw opened the pull request:
Previously puppetdb's generate-string just redirected to
cheshire/generate-string, whereas generate-pretty-string would add
puppetdb-specific date formatting (that rendered seconds at "full"
precision) in addition to setting :pretty to true.

Change generate-string to include the puppetdb-specific date formatting,
since not pretty shouldn't also mean not accurate.

This was causing trouble because information (accuracy) was being lost
in transit (say across http).

One of the tests may fail right now, due to an undesirable interaction between pdb-instance and the log tests, but that appears to be a preexisting issue that can be reproduced via `lein test :only puppetlabs.puppetdb.command-test puppetlabs.puppetdb.testutils.services-test`.  One possible fix: https://github.com/puppetlabs/puppetdb/pull/1567

mullr commented:
lgtm

pljenkinsro commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1703/

senior commented:
@rbrw Seems like a test that ensures the timestamps coming out of the top of PuppetDB includes those the fractional seconds would be good here so that we don't break this in the future. Maybe you could use the specific endpoint that was causing you grief in the test?

rbrw commented:
@senior excellent suggestion.

rbrw commented:
Added the tests, though they cause the existing testutils/services tests to fail.  I investigated and found that the failure could be induced by adding a dummy test file that just tested (with-pdb-instance ... true) that sorted alphabetically before the commant_test.clj (where I added the new test).  This PR provides one way we could fix the problem: https://github.com/puppetlabs/puppetdb/pull/1567

rbrw commented on test/puppetlabs/puppetdb/command_test.clj:
```diff
@@ -1298,6 +1300,45 @@
+                 first
+                 (get "deactivated")
+                 (pt/from-string)))))))
+
```
This is fairly heavyweight - In the longer run, I wonder it might make sense to have a more general roundtrip test (if we don't already) that tests more than just a producer_timestamp against the instance.  Hmm, or would an "echo" service be interesting for general testing of the full encoding/decoding path?

pljenkinsro commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1712/

senior commented:
@rbrw from your comment above, it sounds like we shouldn't merge this PR as it causes testutils/services.clj to fail?

rbrw commented:
@senior actually it turns out that it's is an existing problem on master - both @mullr and @wkalt were able to reproduce it via `lein test :only puppetlabs.puppetdb.command-test puppetlabs.puppetdb.testutils.services-test`, so I think this PR can proceeed.

senior commented on test/puppetlabs/puppetdb/command_test.clj:
```diff
@@ -1298,6 +1300,45 @@
+             (jdbc/with-transacted-connection
+               (:scf-read-db globals)
+               :repeatable-read
+               (from-sql-date (scf-store/node-deactivated-time "foo.local")))))
```
Is there a reason to reach inside PDB and invoke this vs POSTing a deactivated node command?

rbrw replied:
If I understand what you're asking correctly, I just added this because it was easy, and tested exactly what was in the DB without the influence of any extra translations on the way out.  It *might* provide a bit of extra safety, but we don't have to have it.

senior commented on test/puppetlabs/puppetdb/testutils/services.clj:
```diff
@@ -349,3 +350,11 @@
    (fn []
      (pdb-client/submit-command-via-http! base-url cmd version payload))))
+
+(defn wait-for-server-processing [server]
```
I think this needs to have something that fails after some max time.

senior commented on test/puppetlabs/puppetdb/command_test.clj:
```diff
@@ -1298,6 +1300,45 @@
+          ;; enqueue, a DateTime wasn't a problem.
+          input-stamp (java.util.Date. deactivate-ms)
+          expected-stamp (DateTime. deactivate-ms DateTimeZone/UTC)]
+      (enqueue-command connection endpoint
```
Is there a reason to reach inside PDB and invoke this vs POSTing a deactivated node command?

rbrw replied:
Now that enqueue is a method, I was just testing the relevant subsystem more directly, since we want it to work no matter how you reach it, but if we'd prefer to involve the http endpoint, then perhaps the test should also move to http.comand-test.

senior commented on test/puppetlabs/puppetdb/command_test.clj:
```diff
@@ -1298,6 +1300,45 @@
+                              :throw-entire-message true
+                              :query-params {"query"
+                                             (json/generate-string
+                                              ["or" ["=" ["node" "active"] true]
```
Do you need this query clause because you deactivated the node in the previous command?

rbrw replied:
Right - I wanted all the nodes, and was just matching what we do in sync/core.

senior commented on test/puppetlabs/puppetdb/http/events_test.clj:
```diff
@@ -349,7 +349,7 @@
        :containing_class "Foo"
        :report_receive_time "2014-04-16T12:44:40.978Z"
-       :report "99ec099bed6dfb9bff2c7df7828270e95f590147"
+       :report "7d0cfa08901e1e1d80cf2f2f814d356d0e457e09"
```
Did we see the hash change because we were serializing the date to JSON before computing the hash? And thus truncating the milliseconds before computing the hash?

rbrw replied:
I assumed it was just because the rendered formats were slightly different - I think I've seen something like "...8700Z" vs "...087Z"?

rbrw replied:
OK, right, so at least as a practical matter, I added some code to print the json in generate string whenever including the pdb custom date format makes a difference, and saw this:

```
2011-01-01T15:00:01Z
2011-01-01T15:00:01.000Z
```
I was assuming that's what changed the hash, though I haven't investigated the report hashing explicitly.  But I can if you like.

rbrw commented on src/puppetlabs/puppetdb/cheshire.clj:
```diff
@@ -38,19 +38,32 @@
+  ([obj writer]
+   (core/generate-stream obj writer default-opts))
+  ([obj writer opts]
+   (core/generate-stream obj writer (merge default-opts opts))))
```
I overlooked generate-stream in the previous version.

pljenkinsro commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1733/

ajroetker commented on test/puppetlabs/puppetdb/cheshire_test.clj:
```diff
@@ -35,6 +36,14 @@
              "{\n  \"a\" : 1,\n  \"b\" : 2\n}")))))
 
+(deftest test-date-generation
+  (let [t (java.util.Date. 5928174905781)]
```
How come your using the Java interop vs clj-time?

rbrw replied:
This is just checks that our generate-string generates the same (expected) output for both a java.util.Date and a joda DateTime.   Before, Date (specifically), didn't work.  And now that you mention it, perhaps we should check Timestamp here too, since that's one of our important types.  I'll add it in a minute.  You can see the failure by changing generate-string to not include default-opts in the core/generate-string calls.

rbrw commented:
Added a Timestamp test to test-date-generation.

pljenkinsro commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1748/

senior commented on test/puppetlabs/puppetdb/testutils/services.clj:
```diff
@@ -352,3 +355,28 @@
+  "Returns a truthy value indicating whether the wait was
+  successful (i.e. didn't time out).  Given a false or nil timeout,
+  waits forever."
+  ([server] (wait-for-server-processing server nil))
```
Some of this seems unnecessary. I don't see any usage of the 1 arg version. With that gone I think the code below becomes simpler.

rbrw replied:
OK, we were using it before, and I thought I'd preserve it in case someone did want to wait forever, but I suppose they can manage that themselves, or we  can add support later.

senior commented on test/puppetlabs/puppetdb/testutils/services.clj:
```diff
@@ -352,3 +355,28 @@
+         (do
+           (Thread/sleep (if-not deadline
+                           100
+                           (max 0 (min 100 (- deadline (now-ms))))))
```
Any reason why we wouldn't always just sleep 100 then recur? Seems like that would make this simpler

rbrw replied:
Just didn't want to miss the deadline.

rbrw replied:
And I forgot, @mullr asked for this function to be busier in its previous incarnation, so I'll just change it to 10ms (as I did there), and omit the extra work.

pljenkinsro commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1755/
